### PR TITLE
cmake: fix building SDL3 with ndk 26

### PR DESCRIPTION
**Pull request by madebr**
Using the latest Android ndk (26.0.10792818), SDL3 fails to configure and link.
llvm 17 warns about version script entries not defined by the library.

Also, the new Android CMake toolchain file added:
```cmake
if (NOT ANDROID_ALLOW_UNDEFINED_VERSION_SCRIPT_SYMBOLS)
  string(APPEND _ANDROID_NDK_INIT_LDFLAGS " -Wl,--no-undefined-version")
endif()
```
The combination of `-Wl,--no-undefined-version` and `-Wl,--fatal-warnings` causes linking `libSDL3.so` to fail with the following message:
```
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_DYNAPI_entry' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_DXGIGetOutputInfo' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_Direct3D9GetAdapterIndex' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_GDKGetTaskQueue' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_GDKSuspendComplete' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_GetRenderD3D11Device' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_GetRenderD3D9Device' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_LinuxSetThreadPriority' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_LinuxSetThreadPriorityAndPolicy' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_OnApplicationDidChangeStatusBarOrientation' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_RegisterApp' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_RenderGetD3D12Device' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_SetWindowsMessageHook' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_UnregisterApp' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_WinRTGetDeviceFamily' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_WinRTGetFSPathUNICODE' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_WinRTGetFSPathUTF8' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_iPhoneSetAnimationCallback' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_iPhoneSetEventPump' failed: symbol not defined
ld.lld: error: version script assignment of 'SDL3_0.0.0' to symbol 'SDL_GDKGetDefaultUser' failed: symbol not defined
```

This pr fixes these errors by ignoring undefined symbols by adding the `-Wl,--undefined-version` link option.
This solution is **CMake only**. `android.mk` remains broken.

An alternative  solution could be to rewrite the version script to filter out the troublesome symbols by `src/dynapy/gendynapy.py`, and create `src/dynapy/SDL_dynapy_android.sym`. 
`Android.mk` would benefit from this solution as well.

(As a reminder, we're using version scripts on Android because of this [Android middleware advice](https://developer.android.com/ndk/guides/middleware-vendors).

**edit**: a third approach has been implemented: create dummy implementations of _all_ SDL symbols.
This also fixes Android.mk in one go.

### CODE REVIEW

**Comment by slouken** [comment]
Hmm... yeah, the version script shouldn't include symbols which aren't available on the platform. We could alternately implement stubs for those functions which return SDL_Unsupported() on the platforms for which they're not supported.

**Comment by madebr** [comment]
> Hmm... yeah, the version script shouldn't include symbols which aren't available on the platform. We could alternately implement stubs for those functions which return SDL_Unsupported() on the platforms for which they're not supported.

I created stubs in the latest iteration of this pr. Please let me know what you think.

**Comment by slouken** [comment]
> I created stubs in the latest iteration of this pr. Please let me know what you think.

I like it. Thanks!

**Comment by madebr** [comment]
Not 100% sure, but with this change, it is possible to remove `#ifdef`s from the dynapi.
https://github.com/libsdl-org/SDL/blob/a5175e5ed04eb4b4ad1e40dd55be51d7f684af30/src/dynapi/SDL_dynapi_procs.h#L82-118

Looking at the list, I think I also need to add stubs for the android functions.

**Comment by slouken** [comment]
> Not 100% sure, but with this change, it is possible to remove `#ifdef`s from the dynapi.
> 
> https://github.com/libsdl-org/SDL/blob/a5175e5ed04eb4b4ad1e40dd55be51d7f684af30/src/dynapi/SDL_dynapi_procs.h#L82-L118
> 
> Looking at the list, I think I also need to add stubs for the android functions.

Yup, agreed.

**Comment by madebr** [comment]
The latest iterations allows me to build SDL3 with `-DCMAKE_SHARED_LINKER_FLAGS="-Wl,--no-undefined-version"`

(`CMAKE_SHARED_LINKER_FLAGS` is `LDFLAGS` for shared libraries)

File: src/dynapi/SDL_dynapi_procs.h [hunk 1]
```diff
@@ -65,43 +65,30 @@ SDL_DYNAPI_PROC(SDL_Thread*,SDL_CreateThreadWithStackSize,(SDL_ThreadFunction a,
 SDL_DYNAPI_PROC(SDL_Thread*,SDL_CreateThreadWithStackSize,(SDL_ThreadFunction a, const char *b, const size_t c, void *d),(a,b,c,d),return)
 #endif
 
-#if defined(__WIN32__) || defined(__GDK__)
 SDL_DYNAPI_PROC(int,SDL_RegisterApp,(const char *a, Uint32 b, void *c),(a,b,c),return)
 SDL_DYNAPI_PROC(ID3D12Device*,SDL_RenderGetD3D12Device,(SDL_Renderer *a),(a),return)
 SDL_DYNAPI_PROC(void,SDL_SetWindowsMessageHook,(SDL_WindowsMessageHook a, void *b),(a,b),)
 SDL_DYNAPI_PROC(void,SDL_UnregisterApp,(void),(),)
-#endif
```
Comment (madebr) [inline src/dynapi/SDL_dynapi_procs.h:72]: I left the whitespace. Perhaps you want to add some somments?
Also, the new dynapi table will be incompatibe with older (SDL3) ones.

Comment (slouken) [reply src/dynapi/SDL_dynapi_procs.h:72]: We probably don't need comments in SDL_dynapi_procs.h. It's fine for the table to be incompatible, we haven't locked down the ABI yet, and I plan to resort all the functions before we do.

**Comment by slouken** [comment]
Are you waiting on anything before merging this change?

**Comment by madebr** [comment]
No, I'm just letting it "mature" here to make sure interested parties are able to comment.

**Comment by madebr** [comment]
No, I'm just letting it "mature" here to make sure interested parties are able to comment.

**Pull request merged**